src/py/flwr/server/strategy/bulyan_test.py

```diff
@@ -0,0 +1,241 @@
+    """Test num_fit_clients function."""
+    # Prepare
+    strategy = Bulyan()
+    expected = 9
+
+    # Execute
+    actual, _ = strategy.num_fit_clients(num_available_clients=9)
+
+    # Assert
+    assert expected == actual
+
+
+def test_krum_num_evaluation_clients_40_available() -> None:
+    """Test num_evaluation_clients function."""
+    # Prepare
+    strategy = Bulyan(fraction_evaluate=0.05)
+    expected = 2
+
+    # Execute
+    actual, _ = strategy.num_evaluation_clients(num_available_clients=40)
+
+    # Assert
+    assert expected == actual
+
+
+def test_krum_num_evaluation_clients_39_available() -> None:
+    """Test num_evaluation_clients function."""
+    # Prepare
+    strategy = Bulyan(fraction_evaluate=0.05)
+    expected = 2
+
+    # Execute
+    actual, _ = strategy.num_evaluation_clients(num_available_clients=39)
+
+    # Assert
+    assert expected == actual
+
+
+def test_krum_num_evaluation_clients_20_available() -> None:
+    """Test num_evaluation_clients function."""
+    # Prepare
+    strategy = Bulyan(fraction_evaluate=0.05)
+    expected = 2
+
+    # Execute
+    actual, _ = strategy.num_evaluation_clients(num_available_clients=20)
+
+    # Assert
+    assert expected == actual
+
+
+def test_krum_num_evaluation_clients_minimum() -> None:
+    """Test num_evaluation_clients function."""
+    # Prepare
+    strategy = Bulyan(fraction_evaluate=0.05)
+    expected = 2
+
+    # Execute
+    actual, _ = strategy.num_evaluation_clients(num_available_clients=19)
+
+    # Assert
+    assert expected == actual
+
+
+# pylint: disable=too-many-locals
+def test_aggregate_fit() -> None:
+    """Tests if Bulyan is aggregating correctly."""
+    # Prepare
+    previous_weights: NDArrays = [array([0.1, 0.1, 0.1, 0.1], dtype=float32)]
+    strategy = Bulyan(
+        initial_parameters=ndarrays_to_parameters(previous_weights),
+        num_malicious_clients=1,
+    )
+    param_0: Parameters = ndarrays_to_parameters(
+        [array([0.2, 0.2, 0.2, 0.2], dtype=float32)]
+    )
+    param_1: Parameters = ndarrays_to_parameters(
+        [array([0.5, 0.5, 0.5, 0.5], dtype=float32)]
+    )
+    param_2: Parameters = ndarrays_to_parameters(
+        [array([0.7, 0.7, 0.7, 0.7], dtype=float32)]
+    )
+    param_3: Parameters = ndarrays_to_parameters(
+        [array([12.0, 12.0, 12.0, 12.0], dtype=float32)]
+    )
+    param_4: Parameters = ndarrays_to_parameters(
+        [array([0.1, 0.1, 0.1, 0.1], dtype=float32)]
+    )
+    param_5: Parameters = ndarrays_to_parameters(
+        [array([0.1, 0.1, 0.1, 0.1], dtype=float32)]
+    )
+    bridge = MagicMock()
+    client_0 = GrpcClientProxy(cid="0", bridge=bridge)
+    client_1 = GrpcClientProxy(cid="1", bridge=bridge)
+    client_2 = GrpcClientProxy(cid="2", bridge=bridge)
+    client_3 = GrpcClientProxy(cid="3", bridge=bridge)
+    client_4 = GrpcClientProxy(cid="4", bridge=bridge)
+    client_5 = GrpcClientProxy(cid="5", bridge=bridge)
```

Comment:
There's no need to make this test depend on `GrpcClientProxy`. It seems sufficient to use `MagicMock` for the `ClientProxy` instances in the `results` list.